src/psyclone/psyir/backend/sir.py

```diff
@@ -0,0 +1,408 @@
+                  "".format(self._nindent, lhs, rhs))
+        if result[-1] == '\n':
+            # For better formatting, remove the newline if one exists.
+            result = result[:-1]
+        result += ",\n"
+        result += "{0}{1}\"=\"),\n".format(self._nindent, self._indent)
+        return result
+
+    def binaryoperation_node(self, node):
+        '''This method is called when a BinaryOperation instance is found in
+        the PSyIR tree.
+
+        :param node: a BinaryOperation PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.BinaryOperation`
+
+        :returns: the SIR code as a string.
+        :rtype: str
+
+        :raises VisitorError: if there is no mapping from the PSyIR \
+        operator to SIR.
+
+        '''
+        binary_operators = {
+            BinaryOperation.Operator.ADD: '+',
+            BinaryOperation.Operator.SUB: '-',
+            BinaryOperation.Operator.MUL: '*',
+            BinaryOperation.Operator.DIV: '/'}
+
+        self._depth += 1
+        lhs = self._visit(node.children[0])
+        try:
+            oper = binary_operators[node.operator]
+        except KeyError:
+            raise VisitorError(
+                "Method binaryoperation_node in class SIRWriter, unsupported "
+                "operator '{0}' found.".format(str(node.operator)))
+        rhs = self._visit(node.children[1])
+        self._depth -= 1
+        result = "{0}makeBinaryOperator(\n{1}".format(self._nindent, lhs)
+        if result[-1] == '\n':
+            # For better formatting, remove the newline if one exists.
+            result = result[:-1]
+        result += ",\n"
+        result += ("{0}{1}\"{2}\",\n{3}\n{0}{1})\n"
+                   "".format(self._nindent, self._indent, oper, rhs))
+        return result
+
+    def reference_node(self, node):
+        '''This method is called when a Reference instance is found in the
+        PSyIR tree.
+
+        :param node: a Reference PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.Reference`
+
+        :returns: the SIR code as a string.
+        :rtype: str
+
+        :raises VisitorError: if this node has children.
+
+        '''
+        if node.children:
+            raise VisitorError(
+                "Method reference_node in class SIRWriter: SIR Reference "
+                "node is not expected to have any children.")
+        return "{0}makeVarAccessExpr(\"{1}\")".format(self._nindent, node.name)
+
+    def array_node(self, node):
+        '''This method is called when an Array instance is found in the PSyIR
+        tree.
+
+        :param node: an Array PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.Array`
+
+        :returns: the SIR code as a string.
+        :rtype: str
+
+        '''
+        stencil = gen_stencil(node)
+        result = ("{0}makeFieldAccessExpr(\"{1}\",{2})"
+                  "".format(self._nindent, node.name, stencil))
+        # _field_names is a set so duplicates will be ignored.
```

Comment:
Maybe also say what this set will be used for.

Reply:
I repeated what I said in the declaration. Hopefully that is OK.

Reply:
It will be once it has been tweaked c.f. the text for the declaration.

Reply:
Tweaked.